src/ComputeSharp.D2D1/Shaders/Interop/Effects/ResourceManagers/D2D1ResourceTextureManagerImpl.ID2D1ResourceTextureManager.cs

```diff
@@ -0,0 +1,688 @@
+            uint* maximimumExtents,
+            uint* strides,
+            uint dimensions,
+            byte* data,
+            uint dataCount)
+        {
+            return D2D1ResourceTextureManagerImpl.Update(
+                @this,
+                minimumExtents,
+                maximimumExtents,
+                strides,
+                dimensions,
+                data,
+                dataCount);
+        }
+
+        /// <inheritdoc cref="Initialize"/>
+        public static int InitializeWithStagingBuffer(
+            D2D1ResourceTextureManagerImpl* @this,
+            Guid* resourceId,
+            D2D1_RESOURCE_TEXTURE_PROPERTIES* resourceTextureProperties,
+            byte* data,
+            uint* strides,
+            uint dataSize)
+        {
+            // Return E_POINTER if any input pointer is null. In particular, the data pointer
+            // must not be null if the data size is not 0, and the strides pointer must not be
+            // null if some data is supplied, and the number of dimensions is greater than 1.
+            if (resourceTextureProperties is null ||
+                (dataSize > 0 && data is null) ||
+                resourceTextureProperties->extents is null ||
+                resourceTextureProperties->extendModes is null ||
+                (resourceTextureProperties->dimensions > 1 && dataSize > 0 && strides is null))
+            {
+                return E.E_POINTER;
+            }
+
+            // Textures can only have between 1 and 3 dimensions
+            if (resourceTextureProperties->dimensions is 0 or > 3)
+            {
+                return E.E_INVALIDARG;
+            }
+
+            // No extent can be 0
+            foreach (uint extent in new ReadOnlySpan<uint>(resourceTextureProperties->extents, (int)resourceTextureProperties->dimensions))
+            {
+                if (extent == 0)
+                {
+                    return E.E_INVALIDARG;
+                }
+            }
+
+            // The buffer precision can't be unknown, and the channel depth must be explicit
+            if (resourceTextureProperties->bufferPrecision == D2D1_BUFFER_PRECISION.D2D1_BUFFER_PRECISION_UNKNOWN ||
+                resourceTextureProperties->channelDepth == D2D1_CHANNEL_DEPTH.D2D1_CHANNEL_DEPTH_DEFAULT)
+            {
+                return E.E_INVALIDARG;
+            }
+
+            // Validate the data size
+            if (dataSize > 0)
+            {
+                uint elementSizeInBytes = GetElementSizeInBytes(
+                    resourceTextureProperties->bufferPrecision,
+                    resourceTextureProperties->channelDepth);
+
+                int hresult = ValidateDataAndStrides(
+                    elementSizeInBytes,
+                    resourceTextureProperties->extents,
+                    resourceTextureProperties->dimensions,
+                    strides,
+                    dataSize);
+
+                if (hresult != S.S_OK)
```

Comment:
shouldn't this be `if (hresult < 0)` ? do you also return immediately if you get `S_FALSE` ?

Reply:
Well, I wrote that method, and it's only returning either an error or `S_OK`, so it's the same 😄

Reply:
Right but if you change that method you'll have to remember to change this method. Having fewer things to remember is a good thing.